HarmonyOS/OpenHarmony Port

MetsukiMio opened the pull request:
## Description
SDL support for harmonyos

## Existing Issue(s)
#9837

MetsukiMio committed:
Harmony port: Workflows

MetsukiMio committed:
Harmony port: Workflows (disable other plats)

MetsukiMio committed:
Harmony port: Workflows (fix script error)

MetsukiMio committed:
Harmony port: Workflows (fix script error)

MetsukiMio committed:
Harmony port: Workflows (fix script error)

MetsukiMio committed:
Harmony port: window fetching

MetsukiMio committed:
Harmony port: disable lib version suffix

MetsukiMio committed:
Harmony port: video vulkan library & disable version test

MetsukiMio committed:
Harmony port: video vulkan library & disable version test

MetsukiMio committed:
Harmony port: video vulkan library & disable version test

MetsukiMio committed:
Harmony port: video vulkan library

MetsukiMio committed:
Harmony port: video vulkan library

MetsukiMio committed:
Harmony port: video library

MetsukiMio committed:
Harmony port: window creation

Jack253-png commented:
@slouken Request for review

Jack253-png commented:
@madebr @sezero Request for review

MetsukiMio committed:
Harmony port: egl/gles lib

MetsukiMio committed:
Harmony port: keyboard event

MetsukiMio committed:
Harmony port: video drv fix

Jack253-png commented on src/video/SDL_sysvideo.h:
```diff
@@ -542,6 +542,7 @@ extern VideoBootStrap Emscripten_bootstrap;
 extern VideoBootStrap OFFSCREEN_bootstrap;
 extern VideoBootStrap QNX_bootstrap;
 extern VideoBootStrap OPENVR_bootstrap;
+extern VideoBootStrap OHOS_bootstrap;
```
Fixed in the latest commit

MetsukiMio committed:
Harmony port: napi callback shell

madebr commented:
Can you please rebase on top of current master and remove all commits which you did not author?
```sh
git fetch https://github.com/libsdl-org/SDL main
git rebase -i FETCH_HEAD
<remove all commits which you did not author. In vim, "dd" removes the current line, "<ESC>:wq" saves and exits)>
git push git@github.com:Jack253-png/SDL HEAD:OpenMinecraft-Dev -f
```

madebr commented on src/core/ohos/SDL_ohos.c:
```diff
@@ -0,0 +1,246 @@
+SDL_Mutex *g_ohosPageMutex = NULL;
+static OH_NativeXComponent_Callback callback;
+static OH_NativeXComponent_MouseEvent_Callback mouseCallback;
+SDL_WindowData windowData;
+SDL_VideoData videoData;
```
Other platforms usually get `SDL_WindowData` from `SDL_Window.internal`.
`SDL_VideoData` from `SDL_VideoDevice`.

Jack253-png replied:
OHOS only support single window, and the data of the window is fixed, I have to do this

madebr commented on src/core/ohos/SDL_ohos.c:
```diff
@@ -0,0 +1,246 @@
+#include "SDL3/SDL_mutex.h"
+#include "../../video/ohos/SDL_ohoskeyboard.h"
+
+OHNativeWindow *nativeWindow;
+SDL_Mutex *g_ohosPageMutex = NULL;
```
Can these be made static, or member of `VideoDevice`?
If not, perhaps make the name more unique to avoid name clashes?

madebr commented on src/core/ohos/SDL_ohos.c:
```diff
@@ -0,0 +1,246 @@
+static OH_NativeXComponent_MouseEvent_Callback mouseCallback;
+SDL_WindowData windowData;
+SDL_VideoData videoData;
+struct
+{
+    napi_env env;
+    napi_threadsafe_function func;
+    napi_ref interface;
+} napiEnv;
```
Can this struct be static?

madebr commented on src/core/ohos/SDL_ohos.c:
```diff
@@ -0,0 +1,246 @@
+    OH_NativeXComponent_GetXComponentSize(component, window, &width, &height);
+    OH_NativeXComponent_GetXComponentOffset(component, window, &offsetX, &offsetY);
+
+    SDL_Log("Native Window: %p", nativeWindow);
```
Logging priority level is too high. Either use trace level, or remove this.
There are multiple `SDL_Log`s.

madebr commented on src/core/ohos/SDL_ohos.c:
```diff
@@ -0,0 +1,246 @@
+        }
+    }
+}
+static void onNativeTouch(OH_NativeXComponent *component, void *window) {}
+static void onNativeMouse(OH_NativeXComponent *component, void *window) {}
+static void OnDispatchTouchEventCB(OH_NativeXComponent *component, void *window) {}
+void OnHoverEvent(OH_NativeXComponent *component, bool isHover) {}
+void OnFocusEvent(OH_NativeXComponent *component, void *window) {}
+void OnBlurEvent(OH_NativeXComponent *component, void *window) {}
```
Make static, and perhaps add a `// TODO` to make visible what is missing?

madebr commented on src/core/ohos/SDL_ohos.c:
```diff
@@ -0,0 +1,246 @@
+    return exports;
+}
+
+napi_module OHOS_NAPI_Module = {
+    .nm_version = 1,
+    .nm_flags = 0,
+    .nm_filename = NULL,
+    .nm_register_func = SDL_OHOS_NAPI_Init,
+    .nm_modname = "SDL3",
+    .nm_priv = ((void *)0),
+    .reserved = { 0 },
+};
+
+__attribute__((constructor)) void RegisterEntryModule(void)
+{
+    napi_module_register(&OHOS_NAPI_Module);
+}
```
This is not appropriate for within SDL3.
Isn't this something to be put in `SDL3/SDL_main_impl.h`, which users then can include in their main source when doing `#include <SDL3/SDL_main.h>`?

Jack253-png replied:
This is a stub for ohos native module loader, the loader force every loadable library to do this

Jack253-png replied:
And the official document recommends to create a main function(In this case, it's SDL_main) in the library, and runs it in another thread

madebr commented:
Seeing you implemented vulkan support, does SDL_gpu work? Do the [SDL_gpu_examples](https://github.com/TheSpydog/SDL_gpu_examples) work?

MetsukiMio committed:
Harmony port: review changes

MetsukiMio committed:
Harmony port: review changes

Jack253-png commented:
> Seeing you implemented vulkan support, does SDL_gpu work? Do the [SDL_gpu_examples](https://github.com/TheSpydog/SDL_gpu_examples) work?

I will test later, because I don't have test platform, I need to find other people to do this.

MetsukiMio committed:
Harmony port: review changes

MetsukiMio committed:
Harmony port: review changes

MetsukiMio committed:
Harmony port: review changes

Jack253-png commented:
> Seeing you implemented vulkan support, does SDL_gpu work? Do the [SDL_gpu_examples](https://github.com/TheSpydog/SDL_gpu_examples) work?

@StarcloudSea here is the owner of the test platform

MetsukiMio committed:
Harmony port: entrypoint

MetsukiMio committed:
Harmony port: entrypoint

MetsukiMio committed:
Harmony port: entrypoint

MetsukiMio committed:
Harmony port: entrypoint

MetsukiMio committed:
Harmony port: entrypoint

MetsukiMio committed:
Harmony port: entrypoint

Rwagsu committed:
Harmony port: video device

MetsukiMio committed:
Harmony port: call test

MetsukiMio committed:
Harmony port: fix

sezero commented:
Do you need the dynapi functionality of SDL? If you don't, insert a similar disabler for ohos around [here](https://github.com/OpenMinecraft-Dev/SDL/blob/main/src/dynapi/SDL_dynapi.h#L50-L51).

Starcloudsea commented:
To clarify, I'm not a developer, I was just called to test because the computer has good performance, and I don't know about HarmonyOS development. 😅

MetsukiMio committed:
Harmony port: napi shell

Jack253-png commented:
> Do you need the dynapi functionality of SDL? If you don't, insert a similar disabler for ohos around [here](https://github.com/OpenMinecraft-Dev/SDL/blob/main/src/dynapi/SDL_dynapi.h#L50-L51).
> 

Maybe I need it, is it ready for merging?

MetsukiMio committed:
Harmony port: touch event

MetsukiMio committed:
Harmony port: fix

MetsukiMio committed:
Harmony port: fix

MetsukiMio committed:
Harmony port: fix

MetsukiMio committed:
Harmony port: fix

MetsukiMio committed:
Harmony port: surface destroy

MetsukiMio committed:
Harmony port: fix

MetsukiMio committed:
Merge branch 'libsdl-org:main' into main

Rwagsu committed:
Harmony port: fix napi call

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

MetsukiMio committed:
Harmony port: pthread enable

MetsukiMio committed:
Update create-test-plan.py

MetsukiMio committed:
Update generic.yml

MetsukiMio committed:
Update CMakeLists.txt

MetsukiMio committed:
Update create-test-plan.py

MetsukiMio committed:
Update SDL_ohos.c

MetsukiMio committed:
Update CMakeLists.txt

MetsukiMio committed:
Update create-test-plan.py

MetsukiMio committed:
Update create-test-plan.py

MetsukiMio committed:
Harmony port: format

madebr commented on .github/workflows/create-test-plan.py:
```diff
@@ -98,49 +99,51 @@ class JobSpec:
     clang_cl: bool = False
     gdk: bool = False
     vita_gles: Optional[VitaGLES] = None
+    harmony_arch: Optional[str] = None
```
You're not actually using this in the workflow, so I think you can remove this from `create-test-plan`

Jack253-png replied:
No, I planned to do add other two arch

madebr commented on .github/workflows/create-test-plan.py:
```diff
@@ -98,49 +99,51 @@ class JobSpec:
 
 
 JOB_SPECS = {
-    "msys2-mingw32": JobSpec(name="Windows (msys2, mingw32)",               os=JobOs.WindowsLatest,     platform=SdlPlatform.Msys2,       artifact="SDL-mingw32",            msys2_platform=Msys2Platform.Mingw32, ),
```
Please undo these comments, and just make sure to add `[sdl-ci-filter harmony]` in the message of your last commit. This script will then only build haiku and openharmony.

madebr reviewed (commented):
Nice work you did here!

Do all unit tests work? (`ctest` tests with a dummy video driver).
Does running `test/testautomation --video dummy` succeed?
Does running `test/testautomation --video ohos` work?
Does `test/testsprite --renderer opengl` work?
Does `test/testsprite --renderer vulkan` work?
Does `test/testsprite --renderer gpu` work?
Does `test/testfile` work?
Does `test/testfilesystem` work?


Do you also intend to add support for other subsystems?
I looked a bit through the sparse English documentation, and it looks like ohos also has camera support.

Jack253-png commented:
> Nice work you did here!
> 
> Do all unit tests work? (`ctest` tests with a dummy video driver).
> Does running `test/testautomation --video dummy` succeed?
> Does running `test/testautomation --video ohos` work?
> Does `test/testsprite --renderer opengl` work?
> Does `test/testsprite --renderer vulkan` work?
> Does `test/testsprite --renderer gpu` work?
> Does `test/testfile` work?
> Does `test/testfilesystem` work?
> 
> 
> Do you also intend to add support for other subsystems?
> I looked a bit through the sparse English documentation, and it looks like ohos also has camera support.

I need to buy a test platform, I may hang this PR for about 2 weeks, and it is able to create vulkan/egl surface for now (I borrowed another developer's PC and runs the emulator, and it's unavailable for now)

madebr commented:
It would also be useful to write some documentation for the `docs` directory.

Since openharmony is a relatively new os, documentation about how to package a SDL app would be useful. I hope it's not a mess like Android ;)

Jack253-png commented:
> It would also be useful to write some documentation for the `docs` directory.
> 
> Since openharmony is a relatively new os, documentation about how to package a SDL app would be useful. I hope it's not a mess like Android ;)

I have already written a sample harmony/openharmony app at https://github.com/OpenMinecraft-Dev/sdl-ohos-shell, I will try to put it inside

MetsukiMio committed:
Harmony port: workflow fix [sdl-ci-filter harmony]

MetsukiMio committed:
Harmony port: workflow fix [sdl-ci-filter harmony]

sezero commented on .github/workflows/create-test-plan.py:
```diff
@@ -115,7 +117,7 @@ class JobSpec:
     "ubuntu-24.04-arm64": JobSpec(name="Ubuntu 24.04 (ARM64)",              os=JobOs.Ubuntu24_04_arm,   platform=SdlPlatform.Linux,       artifact="SDL-ubuntu24.04-arm64", ),
     "steamrt3": JobSpec(name="Steam Linux Runtime 3.0 (x86_64)",            os=JobOs.UbuntuLatest,      platform=SdlPlatform.Linux,       artifact="SDL-steamrt3",           container="registry.gitlab.steamos.cloud/steamrt/sniper/sdk:latest", ),
-    "steamrt3-arm64": JobSpec(name="Steam Linux Runtime 3.0 (arm64)",       os=JobOs.Ubuntu24_04_arm,   platform=SdlPlatform.Linux,       artifact="SDL-steamrt3-arm64",     container="registry.gitlab.steamos.cloud/steamrt/sniper/sdk/arm64:latest", ),
+    "steamrt3-arm64": JobSpec(name="Steam Linux Runtime 3.0 (arm64)",       os=JobOs.Ubuntu24_04_arm,   platform=SdlPlatform.Linux,       artifact="SDL-steamrt3-arm64",     container="registry.gitlab.steamos.cloud/steamrt/sniper/sdk/arm64:3.0.20250408.124536", ),
```
This change is accidental, I think?

MetsukiMio committed:
Harmony port: workflow fix [sdl-ci-filter harmony]

MetsukiMio committed:
Harmony port: workflow fix [sdl-ci-filter harmony-*]

sezero commented on .github/workflows/create-test-plan.py:
```diff
@@ -143,9 +143,9 @@ class JobSpec:
-    "harmony": JobSpec(name="Harmony",                                      os=JobOs.UbuntuLatest,      platform=SdlPlatform.Harmony,     artifact="SDL-harmony-arm64",     harmony_arch="x86_64"),
+    "harmony": JobSpec(name="Harmony (Arm64)",                              os=JobOs.UbuntuLatest,      platform=SdlPlatform.Harmony,     artifact="SDL-harmony-arm64",     harmony_arch="arm64-v8a"),
+    "harmony": JobSpec(name="Harmony (Arm32)",                              os=JobOs.UbuntuLatest,      platform=SdlPlatform.Harmony,     artifact="SDL-harmony-arm64",     harmony_arch="armeabi-v7a"),
+    "harmony": JobSpec(name="Harmony (x86_64)",                             os=JobOs.UbuntuLatest,      platform=SdlPlatform.Harmony,     artifact="SDL-harmony-arm64",     harmony_arch="x86_64"),
```
If I remember correctly, for the CI filer to work, the filter line must be not in the main sıbject line but it bust be the last line in he commit description. I.e., your commit message should be something like the following:

```
Harmony port: something something

Maybe some other explanation....
[sdl-ci-filter harmony]
```

Jack253-png replied:
I have fixed it, see the latest commit

sezero replied:
doesn't seem to work: it's running all workflows.

Jack253-png replied:
See the latest commit

MetsukiMio committed:
Harmony port: workflow fix [sdl-ci-filter harmony-*]

MetsukiMio committed:
Harmony port: workflow
[sdl-ci-filter harmony-*]

Jack253-png commented:
Does it work? I changed the commit message

sezero commented:
> Does it work? I changed the commit message

Hmph. It still seems to run all the workflows.  @madebr: Is the filtering correct?

madebr commented:
> > Does it work? I changed the commit message
> 
> Hmph. It still seems to run all the workflows. @madebr: Is the filtering correct?

I'm using `${{ github.event.head_commit.message }}` to get the commit message.
It looks like [this works for the run in Jack's fork](https://github.com/OpenMinecraft-Dev/SDL/actions/runs/15513016444), but does not for this PR.

madebr commented:
On pushes, `github.event` is a [push object](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=synchronize#push), which has a `.`head_commit.message` property.

On pull requests, `github.event` is a [`pull_request (synchronize)` object](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=synchronize#pull_request). This objects does not have `.head_commit.message` property. Through `github.event.pull_request.commits_url`, we can find out the commit message.
[This url returns paged data with a maximum of 100 items](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-commits-on-a-pull-request). We need the last item, so it'd require to iterate the list.
The event has `github.event.pull_request.merge_commit_sha`, but does not return the sha of the head commit of the pr. Doing something like `git log -n -1 HEAD~1` does not reliably return the correct commit (remember, we test merge commits in pull requests, not the actual head commit)

Jack253-png commented:
> On pushes, `github.event` is a [push object](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=synchronize#push), which has a `.`head_commit.message` property.
> 
> On pull requests, `github.event` is a [`pull_request (synchronize)` object](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=synchronize#pull_request). This objects does not have `.head_commit.message` property. Through `github.event.pull_request.commits_url`, we can find out the commit message.
> [This url returns paged data with a maximum of 100 items](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-commits-on-a-pull-request). We need the last item, so it'd require to iterate the list.
> The event has `github.event.pull_request.merge_commit_sha`, but does not return the sha of the head commit of the pr. Doing something like `git log -n -1 HEAD~1` does not reliably return the correct commit (remember, we test merge commits in pull requests, not the actual head commit)

It is an existing bug in SDL GitHub workflow?!

madebr commented:
> It is an existing bug in SDL GitHub workflow?!

Yes and no.
Yes because parsing commit messages do not work.
No because pull requests creators cannot sneakily disable a workflow.

sezero commented:
> . No because pull requests creators cannot sneakily disable a workflow.
> 
> So yes, the problem was already present.

Actually that's a good thing and not a problem. A patch causing problems
elsewhere would go undetected otherwise.

MetsukiMio committed:
OpenHarmony Port (renderer)

Jack253-png commented:
@icculus @sezero opengles2 tested, works (but the renderer doesn't work due to the shader problem, I wrote a demo by myself), after solving the multi threading problem, I will put the ohos project into this pr. It is near to the merge

MetsukiMio committed:
Harmony port: entrypoint multithread

Jack253-png commented:
@madebr @sezero request for review

MetsukiMio committed:
fix

MetsukiMio committed:
Harmony port: dialog

MetsukiMio committed:
Harmony port: locale

MetsukiMio committed:
Harmony port: ohos project

madebr commented on ohos-project:
Please no git submodules in SDL.

madebr commented on src/video/SDL_video.c:
```diff
@@ -148,6 +148,9 @@ static VideoBootStrap *bootstrap[] = {
 #endif
 #ifdef SDL_VIDEO_DRIVER_OPENVR
```
Shouldn't the ohos driver have higher priority then the dummy driver?

MetsukiMio committed:
Harmony port: fix

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

Jack253-png commented:
@madebr review

madebr commented:
This looks very great. I've got some questions.
- This adds support for OpenHarmony 5. What compatibility promise does OpenHarmony make? Do apps built for older SDK's run on newer versions?
- Can you document how to package a SDL3 app and create an OpenHarmony app?

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

MetsukiMio committed:
Harmony port: main project

MetsukiMio committed:
Harmony port: doc v0

slouken commented:
@Jack253-png, are you committing to being responsible for the HarmonyOS port? If so, please add some attribution in the README, for example how it was done for the 3DS port:

SDL port for the Nintendo 3DS [Homebrew toolchain](https://devkitpro.org/) contributed by:

-   [Pierre Wendling](https://github.com/FtZPetruska)

Jack253-png commented:
> Do the testautomation tests pass for this platform?

Sorry, for now the official emulator only supports opengles2, so the vulkan/gles3 test cases couldn't pass

Jack253-png commented:
> Can you enable tests in CI, so we can see what's passing?

Unfortunately, it's hard to download emulator in CI (it's terrible internet speed), The opengles2 demo is included, but it's hard for you and other developers outside China to build it.

madebr commented:
> It just crashes with vulkan error -6, it must be tested through the real Huawei phones or tablets

Do you intend to test it on hardware soon?
A simulator is only as good as its model.
Some assurance that SDL3 behaves correctly on the real thing would be nice. (Imho, it is essential)

slouken commented:
Is there anyone with a real Huawei device that can work with you to run tests and validate functionality?

Jack253-png commented:
>  I noticed you added a x86-64 OpenHarmony ci job. Perhaps it is possible to run it on ordinary Intel/AMD hardware?

The official emulator on Windows is x86-64, but the emulator is impossible to embed into the CI testing, it requires a China Huawei account, perhaps I need to find an OpenHarmony emulator for testing. HarmonyOS and OpenHarmony is just like Google Android and AOSP

Jack253-png commented:
But the OpenHarmony is hard to compile, this CI testing maybe impossible if Huawei official developers doesn't notice this library

BA489 commented:
> > I noticed you added a x86-64 OpenHarmony ci job. Perhaps it is possible to run it on ordinary Intel/AMD hardware?
> 
> The official emulator on Windows is x86-64, but the emulator is impossible to embed into the CI testing, it requires a China Huawei account, perhaps I need to find an OpenHarmony emulator for testing. HarmonyOS and OpenHarmony is just like Google Android and AOSP

There is an OpenHarmony-based emulator for global developers in English from European derivative called Oniro that works with QEMU environment, you can check it out: https://docs.oniroproject.org/device-development/developer-boards/emulator/

BA489 commented:
> But the OpenHarmony is hard to compile, this CI testing maybe impossible if Huawei official developers doesn't notice this library

Don't worry I already communicated this with Huawei engineers via Oniro matrix channel

Jack253-png commented:
> > > I noticed you added a x86-64 OpenHarmony ci job. Perhaps it is possible to run it on ordinary Intel/AMD hardware?
> > 
> > The official emulator on Windows is x86-64, but the emulator is impossible to embed into the CI testing, it requires a China Huawei account, perhaps I need to find an OpenHarmony emulator for testing. HarmonyOS and OpenHarmony is just like Google Android and AOSP
> 
> There is an OpenHarmony-based emulator for global developers in English from European derivative called Oniro that works with QEMU environment, you can check it out: https://docs.oniroproject.org/device-development/developer-boards/emulator/ 

Or you can try the current wrapper app, I wrote a opengles2 triangle demo, you could run it

Jack253-png commented:
> > > > > I noticed you added a x86-64 OpenHarmony ci job. Perhaps it is possible to run it on ordinary Intel/AMD hardware?
> > > > 
> > > > 
> > > > The official emulator on Windows is x86-64, but the emulator is impossible to embed into the CI testing, it requires a China Huawei account, perhaps I need to find an OpenHarmony emulator for testing. HarmonyOS and OpenHarmony is just like Google Android and AOSP
> > > 
> > > 
> > > There is an OpenHarmony-based emulator for global developers in English from European derivative called Oniro that works with QEMU environment, you can check it out: https://docs.oniroproject.org/device-development/developer-boards/emulator/
> > 
> > Or you can try the current wrapper app, I wrote a opengles2 triangle demo, you could run it
> 
> perfect! I will test it and keep you updated on it. I want to make sure if you wrote it within SDL or SDL ohos shell itself so I can test it on DevEco Studio with the QEMU Oniro emulator? 
> https://github.com/OpenMinecraft-Dev/SDL
> https://github.com/[OpenMinecraft-Dev/sdl-ohos-shell](https://github.com/OpenMinecraft-Dev/sdl-ohos-shell)

I have already synced it. You can run the one in the SDL repo directly

BA489 commented:
> > > > > > I noticed you added a x86-64 OpenHarmony ci job. Perhaps it is possible to run it on ordinary Intel/AMD hardware?
> > > > > 
> > > > > 
> > > > > The official emulator on Windows is x86-64, but the emulator is impossible to embed into the CI testing, it requires a China Huawei account, perhaps I need to find an OpenHarmony emulator for testing. HarmonyOS and OpenHarmony is just like Google Android and AOSP
> > > > 
> > > > 
> > > > There is an OpenHarmony-based emulator for global developers in English from European derivative called Oniro that works with QEMU environment, you can check it out: https://docs.oniroproject.org/device-development/developer-boards/emulator/
> > > 
> > > 
> > > Or you can try the current wrapper app, I wrote a opengles2 triangle demo, you could run it
> > 
> > 
> > perfect! I will test it and keep you updated on it. I want to make sure if you wrote it within SDL or SDL ohos shell itself so I can test it on DevEco Studio with the QEMU Oniro emulator?
> > https://github.com/OpenMinecraft-Dev/SDL
> > [https://github.com/[OpenMinecraft-Dev/sdl-ohos-shell](https://github.com/OpenMinecraft-Dev/sdl-ohos-shell)](https://github.com/%5BOpenMinecraft-Dev/sdl-ohos-shell%5D(https://github.com/OpenMinecraft-Dev/sdl-ohos-shell))
> 
> I have already synced it. You can run the one in the SDL repo directly

ah great!

BA489 commented:
> > > > > > I noticed you added a x86-64 OpenHarmony ci job. Perhaps it is possible to run it on ordinary Intel/AMD hardware?
> > > > > 
> > > > > 
> > > > > The official emulator on Windows is x86-64, but the emulator is impossible to embed into the CI testing, it requires a China Huawei account, perhaps I need to find an OpenHarmony emulator for testing. HarmonyOS and OpenHarmony is just like Google Android and AOSP
> > > > 
> > > > 
> > > > There is an OpenHarmony-based emulator for global developers in English from European derivative called Oniro that works with QEMU environment, you can check it out: https://docs.oniroproject.org/device-development/developer-boards/emulator/
> > > 
> > > 
> > > Or you can try the current wrapper app, I wrote a opengles2 triangle demo, you could run it
> > 
> > 
> > perfect! I will test it and keep you updated on it. I want to make sure if you wrote it within SDL or SDL ohos shell itself so I can test it on DevEco Studio with the QEMU Oniro emulator?
> > https://github.com/OpenMinecraft-Dev/SDL
> > [https://github.com/[OpenMinecraft-Dev/sdl-ohos-shell](https://github.com/OpenMinecraft-Dev/sdl-ohos-shell)](https://github.com/%5BOpenMinecraft-Dev/sdl-ohos-shell%5D(https://github.com/OpenMinecraft-Dev/sdl-ohos-shell))
> 
> I have already synced it. You can run the one in the SDL repo directly

I done it! Tested it and it runs successfully without any code/runtime errors 
Official SDL graphics library port on HarmonyOS/OpenHarmony work I think
downgraded it to test on API 12 5.0.0
Due to the emulator I have on my machine being HarmonyOS 5.0.0/NEXT API 12 & earlier, pre-API 13 era
https://github.com/user-attachments/assets/83cd5f3c-2593-4941-b734-4b91cc726123

BA489 commented:
> based on opengles2 triangle demo

oh ok, I will research on hooking the audio subsystem of OpenHarmony/HarmonyOS with SDL

Jack253-png commented:
> > based on opengles2 triangle demo
> 
> oh ok, I will research on hooking the audio subsystem of OpenHarmony/HarmonyOS with SDL and test it

Open PR if you have done it

Jack253-png commented:
> > based on opengles2 triangle demo
> 
> oh ok, I will research on hooking the audio subsystem of OpenHarmony/HarmonyOS with SDL and test it

Any progress?

MetsukiMio committed:
Merge branch 'libsdl-org:main' into main

Jack253-png commented:
> > based on opengles2 triangle demo
> 
> oh ok, I will research on hooking the audio subsystem of OpenHarmony/HarmonyOS with SDL and test it

Have you succeed in porting audio subsystem?

MetsukiMio committed:
OpenHarmony: locale subsystem

MetsukiMio committed:
Merge branch 'main' of https://github.com/libsdl-org/SDL

MetsukiMio committed:
OpenHarmony: shell update

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

MetsukiMio committed:
Merge branch 'libsdl-org:main' into main

MetsukiMio committed:
OpenHarmony port: events

MetsukiMio committed:
OpenHarmony port: fix

MetsukiMio committed:
Merge branch 'libsdl-org:main' into main

MetsukiMio committed:
OpenHarmony: fix

MetsukiMio committed:
Merge branch 'main' of github.com:OpenMinecraft-Dev/SDL

MetsukiMio committed:
OpenHarmony: fix

MetsukiMio committed:
Merge branch 'libsdl-org:main' into main

MetsukiMio committed:
OpenHarmony: test web view

MetsukiMio committed:
Merge branch 'main' of github.com:OpenMinecraft-Dev/SDL

MetsukiMio committed:
Update SDL_ohosvideo.c

MetsukiMio committed:
OpenHarmony: test

MetsukiMio committed:
OpenHarmony: misc api

MetsukiMio committed:
OpenHarmony: battery API

MetsukiMio committed:
OpenHarmony: file selection

MetsukiMio committed:
OpenHarmony: dialog api finish

MetsukiMio committed:
OpenHarmony: sensors

MetsukiMio committed:
Merge branch 'main' of https://gh-proxy.com/https://github.com/libsdl-org/SDL

MetsukiMio committed:
OpenHarmony: dialog / sensor / filesystem subsystem

MetsukiMio committed:
Merge remote-tracking branch 'origin/main'

MetsukiMio committed:
Update action.yml

bytemain committed:
fix(ohos): process only the changed touch point in touch events

The previous implementation processed all touch points in the array for every touch event, which could cause duplicate or incorrect events since the array contains the current state of all fingers. Now we correctly identify and process only the finger that triggered the current event (touchEvent.id), improving touch event accuracy on OpenHarmony platforms.

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

MetsukiMio committed:
Merge pull request #1 from bytemain/main

OHOS: Fix touch event processing to only handle the triggering finger

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

MetsukiMio committed:
OpenHarmony: merge / fix

MetsukiMio committed:
Merge remote-tracking branch 'origin/main'

MetsukiMio committed:
OpenHarmony: port & fix to 6.0.0

MetsukiMio committed:
OpenHarmony: more events

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

MetsukiMio committed:
OpenHarmony: text edit

xiangjs6 commented:
我在鸿蒙笔记本上使用cmake 目前编译会报错，暂时无法提供报错信息，我在华为店的笔记本上尝试的

MetsukiMio commented:
> 我在鸿蒙笔记本上使用cmake 目前编译会报错，暂时无法提供报错信息，我在华为店的笔记本上尝试的

你在自己的deveco studio上试试（

MetsukiMio committed:
Merge branch 'main' of github.com:libsdl-org/SDL

MetsukiMio committed:
openharmony: update & merge

MetsukiMio committed:
openharmony: fix script

MetsukiMio committed:
openharmony: fix

MetsukiMio committed:
openharmony: fix

MetsukiMio committed:
openharmony: fix

sezero commented on CMakeLists.txt:
```diff
@@ -1709,6 +1709,84 @@ if(ANDROID)
     target_link_options(SDL3-static INTERFACE "-Wl,-u,JNI_OnLoad")
   endif()
 
+if(TARGET SDL3-shared)
+  target_link_options(SDL3-shared PRIVATE "-Wl,-z,max-page-size=16384")
+  target_link_options(SDL3-shared PRIVATE "-Wl,-z,common-page-size=16384")
+endif()
```
Why are these  not in the `elseif(OHOS)` case below?

MetsukiMio replied:
These configs are not for ohos

sezero replied:
A bad merge then? Because you are adding those lines to the android case.

sezero replied:
These wrong additions still aren't removed.

sezero commented on ohos-project/entry/src/main/cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+# the minimum version of CMake.
+cmake_minimum_required(VERSION 3.5.0)
```
This should read `cmake_minimum_required(VERSION 3.5...4.0)` in order to avoid cmake deprecation warnings.

MetsukiMio replied:
HarmonyOS toolchain provides a old version of cmake, so I have to do this

sezero replied:
The `...4.0` part will not force you to use a newer cmake: https://cmake.org/cmake/help/v4.0/command/cmake_minimum_required.html

MetsukiMio committed:
merge upstream

MetsukiMio committed:
ohos: update

MetsukiMio committed:
ohos: fix

MetsukiMio committed:
Merge branch 'main' of https://github.com/libsdl-org/SDL

MetsukiMio committed:
openharmony: fix

MetsukiMio committed:
merge upstream

icculus commented:
I should warn you that I have a separate port to HarmonyOS in a fork, and it's pretty far along: https://github.com/icculus/SDL/tree/sdl3-harmonyos

MetsukiMio commented:
> I should warn you that I have a separate port to HarmonyOS in a fork, and it's pretty far along: https://github.com/icculus/SDL/tree/sdl3-harmonyos
> 
> 

so this branch is nearly complete？

MetsukiMio commented:
> I should warn you that I have a separate port to HarmonyOS in a fork, and it's pretty far along: https://github.com/icculus/SDL/tree/sdl3-harmonyos
> 
> 

The locale api is missing

MetsukiMio commented:
> I should warn you that I have a separate port to HarmonyOS in a fork, and it's pretty far along: https://github.com/icculus/SDL/tree/sdl3-harmonyos
> 
> 

Ok, I see. I will try this branch later, and I may try to port my implementation of locale subsystem, Does it have a simple renderer test?

icculus commented:
I've been adding files from the test directory to the openharmony-project directory to try them, but I haven't automated this yet.

MetsukiMio commented:
> I've been adding files from the test directory to the openharmony-project directory to try them, but I haven't automated this yet.

sorry，It didn't work

MetsukiMio commented:
> I've been adding files from the test directory to the openharmony-project directory to try them, but I haven't automated this yet.

make the demo project renders a triangle, and I meet some problem during the test (my simulator seems unable to use Vulkan or glesv3)

icculus commented:
Vulkan and GLES3 are definitely working here.

https://github.com/user-attachments/assets/44df2cc2-9c16-4c84-a1cd-a55635038ce4

(That video is using GLES3 through SDL's Render API, but testvulkan worked here, as does testgpu_spinning_cube.)

MetsukiMio commented:
> Vulkan and GLES3 are definitely working here.
> 
> https://github.com/user-attachments/assets/44df2cc2-9c16-4c84-a1cd-a55635038ce4
> 
> (That video is using GLES3 through SDL's Render API, but testvulkan worked here, as does testgpu_spinning_cube.)

Maybe my simulator have this problem, and could you try to make it work on GLES2?

MetsukiMio commented:
> Vulkan and GLES3 are definitely working here.
> 
> https://github.com/user-attachments/assets/44df2cc2-9c16-4c84-a1cd-a55635038ce4
> 
> (That video is using GLES3 through SDL's Render API, but testvulkan worked here, as does testgpu_spinning_cube.)

So how to run the tests?

MetsukiMio commented:
Maybe I need a raw opengles2/3 demo

MetsukiMio commented:
And probably 2in1, tablet devices are also needed(mouse events, etc.)，simulator tests are essential for these devices

MetsukiMio commented:
And the raw main() function entrypoint is not supported, the app just crashed immediately

MetsukiMio commented:
I tried the opengles rect demo, and it stucked on a black screen

MetsukiMio committed:
merge upstream
